### PR TITLE
SSO with SAML

### PR DESCRIPTION
**Pull request by kitallis**
**Closes:** #605 

Screenshot 2024-07-17 at 5 35 51 PM

Primary things still pending:

- [x] Alternative invite flow for SSO (tinier)
- [x] Test that  EmailAuthentication invite flow continues to work (after the refactor)
- [x] Organization settings page
- [x] Ensure admin account works
- [x] Ensure no unnecessary DB calls is being made around SSO for EmailAuthentication flow
- [x] Handle/catch Descope-related errors a bit better. Rather than general catch-all
- [x] Test with an Azure AD account (full flow with configuration link)

### CODE REVIEW

File: app/controllers/application_controller.rb [hunk 1]
```diff
@@ -46,20 +66,31 @@ def n_plus_one_detection
...
+      !authentication_controllers? &&
       !request.xhr? &&
       !turbo_frame_request?
   end
 
   def store_user_location!
     store_location_for(:user, request.fullpath)
   end
+
+  def after_sign_in_path_for(_)
```
Comment (kitallis) [inline app/controllers/application_controller.rb:85]: 📝

This is overriden on a top-level and the argument is rejected because we'll always store the resource name as `:user` rather than `:email_authentication` or `:sso_authentication` for simplicity.

File: app/controllers/authentication/sessions_controller.rb [hunk 2]
```diff
@@ -0,0 +1,12 @@
+class Authentication::SessionsController < ApplicationController
```
Comment (kitallis) [inline app/controllers/authentication/sessions_controller.rb:1]: 📝

This is a sort of jump controller that decides where to take the user. Adjacently, the `authenticated` route from devise in `routes.rb` has been removed and this general root handler is now what decides.

File: app/controllers/concerns/authenticatable.rb [hunk 3]
```diff
@@ -0,0 +1,51 @@
+module Authenticatable
```
Comment (kitallis) [inline app/controllers/concerns/authenticatable.rb:1]: 📝

This is a general AuthN helper but mostly deals with SSO for now since Devise internally handles the rest. 

The `skip_authentication` is the only general method right now.

File: app/controllers/concerns/authenticatable.rb [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+  SSO_JWT_SESSION_KEY = :sso_session
+  SSO_JWT_REFRESH_KEY = :sso_refresh
+  SSO_JWT_USER_ID_KEY = :sso_user_id
+
+  def skip_authentication
```
Comment (kitallis) [inline app/controllers/concerns/authenticatable.rb:6]: 📝

This method is similar in principle to [require_no_authentication](https://github.com/heartcombo/devise/blob/a259ff3c28912a27329727f4a3c2623d3f5cb6f2/app/controllers/devise_controller.rb#L116).

It short-circuits auth/session related pages and redirects users to the primary logged in page.

`require_no_authentication` is insufficient when you have the following combinations:

logged in via email, visit sso session pages
logged in via sso, visit email session pages
logged in via email, visit email session pages
logged in via sso, visit sso session pages

Basically, if you're logged in as any other login method, and if you visit any kind of login-auth pages, short-circuit.

File: app/controllers/signed_in_application_controller.rb [hunk 5]
```diff
@@ -1,31 +1,43 @@
...
   rescue_from NotAuthorizedError, with: :user_not_authorized
 
   protected
 
-  helper_method :demo_org?, :demo_train?, :subscribed_org?, :billing?, :billing_link, :logout_path
-  PATH_PARAMS_UNDER_APP = [:id, :app_id, :integration_id, :train_id, :platform_id]
-
   def logout_path
-    destroy_email_authentication_session_path
+    if login_by_sso?
```
Comment (kitallis) [inline app/controllers/signed_in_application_controller.rb:36]: 📝

This is the important change here.

File: app/controllers/signed_in_application_controller.rb [hunk 6]
```diff
@@ -1,31 +1,43 @@
 class SignedInApplicationController < ApplicationController
+  include Authenticatable
   DEFAULT_TIMEZONE = "Asia/Kolkata"
-  include MetadataAwareness
+  DEFAULT_TIMEZONE_LIST_REGEX = /Asia\/Kolkata/
+  PATH_PARAMS_UNDER_APP = [:id, :app_id, :integration_id, :train_id, :platform_id]
+
+  layout -> { ensure_supported_layout("signed_in_application") }
 
+  before_action :authenticate_sso_request!, if: :login_by_sso?
```
Comment (kitallis) [inline app/controllers/signed_in_application_controller.rb:9]: 📝

This is the important change here.

File: app/controllers/concerns/metadata_awareness.rb
Comment (kitallis) [inline app/controllers/concerns/metadata_awareness.rb:1]: 📝

This file is deleted because its methods have moved to `ApplicationController`. 

On digging further, Devise eventually hands off the inheritance to `ApplicationController` so this is not really useful. So the `current_*` methods are easily already available to any AuthN strategy that inherits from `ApplicationController`.

File: app/controllers/concerns/authenticatable.rb [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+    end
+  end
+
+  def authenticate_sso_request!
+    return unless login_by_sso?
+
+    st, rt = session[SSO_JWT_SESSION_KEY], session[SSO_JWT_REFRESH_KEY]
+    result = Accounts::SsoAuthentication.validate_or_refresh_session(st, rt)
+
+    # TODO: path where user has externally logged out, log them out of here as well
```
Comment (kitallis) [inline app/controllers/concerns/authenticatable.rb:24]: Even though this TODO exists, I think this is already handled because of [this](https://github.com/tramlinehq/tramline/blob/df6a8a953990fb606bb03db182500eb3b6064b4a/app/controllers/signed_in_application_controller.rb#L56). Unconfirmed.

File: app/models/accounts/user_authentication.rb [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+#  created_at           :datetime         not null
+#  updated_at           :datetime         not null
+#  authenticatable_id   :uuid             not null, indexed => [authenticatable_type]
+#  user_id              :uuid             indexed
+#
+class Accounts::UserAuthentication < ApplicationRecord
+  belongs_to :user
+  belongs_to :authenticatable, polymorphic: true
+
+  validates :authenticatable_id, uniqueness: {scope: [:user_id, :authenticatable_type]}
```
Comment (kitallis) [inline app/models/accounts/user_authentication.rb:16]: Across organizations?
